### PR TITLE
Feat/sma 731 foundry test

### PR DESCRIPTION
**Pull request by Aboudjem**

### CODE REVIEW

File: contracts/base/ERC4337Account.sol [hunk 1]
```diff
@@ -51,6 +51,7 @@ abstract contract ERC4337Account is IERC4337Account {
 
     function addDeposit() public payable virtual {
         IEntryPoint(entryPoint()).depositTo{ value: msg.value }(address(this));
+        emit DepositAdded(address(this), msg.sender, msg.value);
```
Comment (livingrockrises) [inline contracts/base/ERC4337Account.sol:54]: can change to some other name

besides entrypoint already emits this event
emit Deposited(account, newDeposit);

File: contracts/interfaces/IERC4337Account.sol [hunk 2]
```diff
@@ -4,6 +4,8 @@ pragma solidity ^0.8.24;
 import { PackedUserOperation } from "account-abstraction/contracts/interfaces/PackedUserOperation.sol";
 
 interface IERC4337Account {
+    event DepositAdded(address indexed account, address indexed depositor, uint256 amount);
```
Comment (livingrockrises) [inline contracts/interfaces/IERC4337Account.sol:7]: all three can be indexed

File: test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteBatch.t.sol [hunk 3]
```diff
@@ -65,8 +65,34 @@ contract TestAccountExecution_ExecuteBatch is TestAccountExecution_Base {
...
 
         ENTRYPOINT.handleOps(userOps, payable(BOB.addr));
     }
+
+    function test_ExecuteBatch_ValueTransfer() public {
+        address receiver = address(0x123);
+        uint256 sendValue = 1 ether;
+
+        // Fund BOB_ACCOUNT with 10 ETH to cover the value transfer
+        payable(address(BOB_ACCOUNT)).call{ value: 10 ether }(""); // Fund BOB_ACCOUNT
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteBatch.t.sol:78]: cc @filmakarov 
do we need to test it here with supplying 2300 gas?

File: test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteFromExecutor.t.sol [hunk 4]
```diff
@@ -20,8 +20,13 @@ contract TestAccountExecution_ExecuteFromExecutor is Test, SmartAccountTestLab {
         // Install MockExecutor as executor module on BOB_ACCOUNT
         bytes memory callDataInstall =
             abi.encodeWithSelector(IModuleManager.installModule.selector, uint256(2), address(mockExecutor), "");
-        PackedUserOperation[] memory userOpsInstall = prepareExecutionUserOp(
-            BOB, BOB_ACCOUNT, EXECTYPE_DEFAULT, address(BOB_ACCOUNT), 0, callDataInstall
+        
+        
+        Execution[] memory execution = new Execution[](1);
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteFromExecutor.t.sol:25]: now that we have Execution array
type of execution target, value and data can be prepared by separate helpers.
erc20 , native transfer, mock protocol, counter etc for example 
more helpers can be created in same fashion to install a module, to make install data for particular module
and we just use those utils to make execution array element

Comment (livingrockrises) [reply test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteFromExecutor.t.sol:25]: instead of doing something like this 

execution[0] = Execution(address(mockExecutor), 0, execCallData);

enconding data for execCallData can have separate sub utils.

or entire util function for preparing specific nature of **Execution**

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+        account.addDeposit{value: depositAmount}();
+
+        assertEq(balanceBefore + depositAmount, ENTRYPOINT.balanceOf(address(account)), "Deposit should be added to EntryPoint");
+    }
+
+    function test_AddDeposit_EventEmitted() public {
+        uint256 depositAmount = 1 ether;
+        (bool res, ) = address(account).call{value: depositAmount}(""); // Pre-funding the account contract
+        assertTrue(res, "Pre-funding account should succeed");
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:29]: what does it mean "pre-funding"
would this mean undeployed account counterfactual?

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+function test_AddDeposit_DepositViaHandleOps() public {
+    (bool res, ) = payable(address(BOB_ACCOUNT)).call{ value: 2 ether }(""); // Fund account with 2 ETH
+    assertTrue(res, "Pre-funding account should succeed");
+    uint256 depositAmount = 1 ether;
+    uint balanceBefore = ENTRYPOINT.balanceOf(address(account));
+
+    Execution[] memory execution = new Execution[](1);
+    execution[0] = Execution(address(account), depositAmount, abi.encodeWithSignature("addDeposit()"));
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:48]: this case doesn't make sense to do it with user operation? account depositing into entrypoint from native tokens deposited already within itself?
because generally our modifier addMissingFunds takes care of it...

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        EXECTYPE_DEFAULT,
+        executions
+    );
+        uint256 gasStart = gasleft();
+    ENTRYPOINT.handleOps(userOps, payable(BOB.addr));
+    uint gasUsed = gasStart - gasleft();
+
+    // Using almostEq to compare balances with a tolerance for gas costs
+    uint256 maxPercentDelta = 100_000_000_000;
+    almostEq(balanceBefore + (depositAmount * 5)  - (gasUsed * tx.gasprice), ENTRYPOINT.balanceOf(address(account)), maxPercentDelta);
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:92]: can do this test with paymaster involved, then it should be equal

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+function test_AddDeposit_Try_DepositViaHandleOps() public {
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:95]: generally depositing via handleOps is not a use case 
so these tests don't add much value here.
for coverage you can just make 
i. eoas call addDeposit
ii. have account funded with enough native and making sure modifier addMissingFunds does its job internally automatically
iii. no need for such batch tests here. we can put this time into writing other tests

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_EntryPoint.t.sol [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+import "../../../utils/SmartAccountTestLab.t.sol";
+
+contract TestERC4337Account_EntryPoint is Test, SmartAccountTestLab {
+
+    function setUp() public {
+        init();
+    }
+
+    function test_CorrectEntryPointAddress() public {
+        assertEq(BOB_ACCOUNT.entryPoint(), address(ENTRYPOINT), "Should return the correct EntryPoint address");
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_EntryPoint.t.sol:14]: we also want to cover tests of being able to query implementation address, account version (if to be defined) etc
cc @filmakarov

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_GetDeposit.t.sol [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    function test_DepositAfterAddDepositCall() public {
+        uint balanceBefore = ENTRYPOINT.balanceOf(address(BOB_ACCOUNT));
+        uint256 depositAmount = 2 ether;
+        payable(address(BOB_ACCOUNT)).call{value: 1.5 ether}(""); // Sending ether to the account contract directly
+        BOB_ACCOUNT.addDeposit{value: depositAmount}(); // Function that triggers a deposit to the EntryPoint
+
+        // Using almostEq to compare balances with a tolerance for gas costs
+        uint256 maxPercentDelta = 100_000_000_000;
+        almostEq(balanceBefore + depositAmount, ENTRYPOINT.balanceOf(address(BOB_ACCOUNT)), maxPercentDelta);
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_GetDeposit.t.sol:28]: this should be exactly equal as you didt't send any userop

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_OnlyEntryPointOrSelf.t.sol [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    MockValidator public validator;
+    address public userAddress;
+
+    function setUp() public {
+        init();
+        userAddress = address(BOB.addr);
+        validator = new MockValidator();
+    }
+
+    function test_ValidateUserOp_ValidOperation() public {
```
Comment (livingrockrises) [inline test/foundry/unit/concrete/erc4337account/TestERC4337Account_OnlyEntryPointOrSelf.t.sol:19]: these tests are not really related to access control onlyEntryPointOrSelf but sig authorization

**Review by livingrockrises — COMMENTED** [review]
reviewed, looks good.
check comments
do start adding tests for different kinds of transactions which will be useful for gas benchmarking.
also make sure you leave a room for functions like prepareUserOperation to be able to supply paymaster later on, also being able to do transactions on undeployed account and also on deployed account.
this is why we need utils to just get counterfactual address and initcode for undeployed account and pass it in userop building methods..

File: test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteBatch.t.sol [hunk 12]
```diff
@@ -17,7 +17,7 @@ contract TestAccountExecution_ExecuteBatch is TestAccountExecution_Base {
         executions[1] = Execution(address(counter), 0, abi.encodeWithSelector(Counter.incrementNumber.selector));
 
         // Execute batch operation
-        PackedUserOperation[] memory userOps = prepareBatchExecutionUserOp(BOB, BOB_ACCOUNT, EXECTYPE_DEFAULT, executions);
+        PackedUserOperation[] memory userOps = prepareUserOperation(BOB, BOB_ACCOUNT, EXECTYPE_DEFAULT, executions);
```
[the remark this answers is not in the document]
Comment (Aboudjem) [reply test/foundry/unit/concrete/accountexecution/TestAccountExecution_ExecuteBatch.t.sol:20]: yes signer then sender (SA)

File: contracts/interfaces/IERC4337Account.sol (hunk 2 above)
Comment (Aboudjem) [reply contracts/interfaces/IERC4337Account.sol:7]: that's really a bad practice to index amount.

File: contracts/base/ERC4337Account.sol (hunk 1 above)
Comment (Aboudjem) [reply contracts/base/ERC4337Account.sol:54]: ok I'll remove it, if not necessary

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol (hunk 6 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:48]: It's unit testing, checking all cases, even the obvious, also try to reproduce all case through handleOps flow.

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol (hunk 7 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:92]: Yes, can add additional tests with a MockPaymaster

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol (hunk 8 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_AddDeposit.t.sol:95]: it's unit testing, so trying all cases, and edge cases to make sure that they are no unexpected errors. Another Account could be calling addDeposit on another account. 

It can involve two Accounts here if that makes it easier to understand.

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_EntryPoint.t.sol (hunk 9 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_EntryPoint.t.sol:14]: not sure to understand this, TestAccoungConfig_AccountId is already testing it.

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_GetDeposit.t.sol (hunk 10 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_GetDeposit.t.sol:28]: It's not exactly the same number as `deployAccount` on Helpers is calling handleOps once to deploy the smart Account contract

File: test/foundry/unit/concrete/erc4337account/TestERC4337Account_OnlyEntryPointOrSelf.t.sol (hunk 11 above)
Comment (Aboudjem) [reply test/foundry/unit/concrete/erc4337account/TestERC4337Account_OnlyEntryPointOrSelf.t.sol:19]: yes actually, that's a mistake, I duplicated `TestERC4337Account_ValidateUserOp.t.sol` to edit it, It was not supposed to be pushed 😅
